src/StructTypes.jl

```diff
@@ -103,6 +105,20 @@ StructTypes.StructType(::Type{CoolType}) = StructTypes.Struct()
 """
 abstract type Struct <: DataType end
 
+
+# Check for structs we can't use multiple dispatch for
+function StructType(::Type{T}) where {T}
+    if T <: Core.Function
+        NoStructType()
+    elseif Base.issingletontype(T)
+        SingletonType()
+    elseif Base.isabstracttype(T)
+        AbstractType()
```

Comment:
I don't think this is a good idea; the `AbstractType` in StructTypes.jl has a more specific meaning and usage compared with an abstract _julia_ type. Take `Signed` or `Real` for example, those aren't `AbstractType` in the StructTypes.jl sense; they're `NumberType`s. I think in lots of cases people will define a whole tree of types as `Struct` or otherwise.

I also thought the motivation for this PR was making `Struct` the default? I was expecting the `else` clause below to return `Struct` in the non-exceptional case.

Reply:
My idea for this MR comes from this comment: https://github.com/JuliaData/StructTypes.jl/pull/67#issuecomment-983822907

Specifically
> (interesting side effects when things aren't structs) was the original motivation for not making the default Struct.

I figured that Every type that could be defined is either subtyped from `Base`, subtyped from `Core`, or created in one of the ways I added tests for [here](https://github.com/JuliaData/StructTypes.jl/pull/70/files#diff-3b9314a6f9f2d7eec1d0ef69fa76cfabafdbe6d0df923768f9ec32f27a249c63R256).  With that idea I'm hoping to handle every type of possible struct in a sensible default way

I'm not sure that's possible to do in a sensible way though which is why I posted this as a draft MR and linked it in the "default to `struct`" issue

Here I specifically considered types like `Number` where it's abstract but all it's concrete subtypes should be `NumberType` or something more specific.  It does cause some strangeness in the tests for the Core and Base subtypes where abstract types appear twice in the test data, like where `Number` appears [here](https://github.com/JuliaData/StructTypes.jl/pull/70/files#diff-3b9314a6f9f2d7eec1d0ef69fa76cfabafdbe6d0df923768f9ec32f27a249c63R26) and [here](https://github.com/JuliaData/StructTypes.jl/pull/70/files#diff-3b9314a6f9f2d7eec1d0ef69fa76cfabafdbe6d0df923768f9ec32f27a249c63R47).  The idea is that if you get a type with `typeof` it should always be concrete, and I assume that if a user is querying classes directly they want to work with mixed types under that class or they're querying that type itself (like if they use `typejoin` or `typeintersect`.  I'm totally fine with removing that if that edge case isn't an issue though

Reply:
I removed the `isabstracttype` check